radix-engine/src/engine/modules/execution_trace.rs

```diff
@@ -249,15 +137,127 @@ impl<R: FeeReserve> Module<R> for ExecutionTraceModule {
         self.handle_post_sys_call(call_frame, heap, output)
     }
 
-    fn on_run(
+    fn pre_execute_invocation(
         &mut self,
         actor: &REActor,
         input: &IndexedScryptoValue,
         call_frame: &CallFrame,
         heap: &mut Heap,
         track: &mut Track<R>,
     ) -> Result<(), ModuleError> {
-        Self::trace_run(call_frame, heap, track, actor, input);
+        if self.current_sys_call_depth <= self.max_sys_call_trace_depth {
+            let origin = match actor {
+                REActor::Method(
+                    ResolvedMethod::Scrypto {
+                        package_address,
+                        blueprint_name,
+                        ident,
+                        ..
+                    },
+                    ..,
+                ) => SysCallTraceOrigin::ScryptoMethod(
+                    *package_address,
+                    blueprint_name.clone(),
+                    ident.clone(),
+                ),
+                REActor::Function(ResolvedFunction::Scrypto {
+                    package_address,
+                    blueprint_name,
+                    ident,
+                    ..
+                }) => SysCallTraceOrigin::ScryptoFunction(
+                    *package_address,
+                    blueprint_name.clone(),
+                    ident.clone(),
+                ),
+                REActor::Method(ResolvedMethod::Native(native_method), ..) => {
+                    SysCallTraceOrigin::NativeMethod(native_method.clone())
+                }
+                REActor::Function(ResolvedFunction::Native(native_function)) => {
+                    SysCallTraceOrigin::NativeFunction(native_function.clone())
+                }
+            };
+
+            let trace_data = Self::extract_trace_data(heap, input)?;
+            self.traced_sys_call_inputs_stack.push((
+                trace_data,
+                origin,
+                self.current_instruction_index,
+            ));
+        }
+
+        self.current_sys_call_depth += 1;
+        if let REActor::Method(ResolvedMethod::Native(native_method), resolved_receiver) = actor {
+            let caller = &call_frame.actor;
+
+            match (native_method, resolved_receiver.receiver) {
+                (NativeMethod::Vault(VaultMethod::Put), RENodeId::Vault(vault_id)) => {
+                    Self::handle_vault_put(heap, track, caller, &vault_id, input)
+                }
+                (NativeMethod::Vault(VaultMethod::Take), RENodeId::Vault(vault_id)) => {
+                    Self::handle_vault_take(track, caller, &vault_id, input)
+                }
+                (NativeMethod::Vault(VaultMethod::LockFee), RENodeId::Vault(vault_id)) => {
+                    Self::handle_vault_lock_fee(track, caller, &vault_id)
+                }
+                _ => {}
+            }
+        }
+
+        Ok(())
+    }
+
+    fn post_execute_invocation(
```

Comment:
This method duplicates a lot of code from `handle_post_sys_call`. Wdyt about cleaning it up a bit and extracting common code into methods?

Reply:
Yup sounds good

Reply:
Seems like this may be harder than it looks without making code even more convoluted. I'd suggest we keep code as is and then add comments that we still need to find correct abstraction. How does that sound?

Reply:
That sounds like a veiled way to say "let's push this problem to the next person changing this file" :D

Reply:
I removed some duplication in my commit (please review), but it still isn't ideal.
It's quite tricky to follow the internal state (specifically `current_sys_call_depth` and the inputs stack) - it's now being updated in two distinct places. I understand this is not a priority though so I'm okay to keep it as is for now, and perhaps add some explanatory comments to make that easier to grasp for our future selves.

Reply:
Yes :) My point is that extracting common code based on just "pattern matching" (rather than correct abstraction) may actually lead to the wrong abstractions and make later refactors more difficult than necessary.

So imo better to have "messy, duplicated code with comments" rather than "de-duped but incorrectly abstracted code".